src/pkg/podman/podman.go

```diff
@@ -0,0 +1,317 @@
+	}
+
+	return info[0], nil
+}
+
+// GetContainers is a wrapper function around `podman ps --format json` command.
+//
+// Parameter args accepts an array of strings to be passed to the wrapped command (eg. ["-a", "--filter", "123"]).
+//
+// Returned value is a slice of dynamically unmarshalled json, so it needs to be treated properly.
+//
+// If a problem happens during execution, first argument is nil and second argument holds the error message.
+func GetContainers(args ...string) ([]map[string]interface{}, error) {
+	args = append([]string{"ps", "--format", "json"}, args...)
+	output, err := CmdOutput(args...)
+	if err != nil {
+		return nil, err
+	}
+
+	var containers []map[string]interface{}
+
+	err = json.Unmarshal(output, &containers)
+	if err != nil {
+		return nil, err
+	}
+
+	return containers, nil
+}
+
+// GetImages is a wrapper function around `podman images --format json` command.
+//
+// Parameter args accepts an array of strings to be passed to the wrapped command (eg. ["-a", "--filter", "123"]).
+//
+// Returned value is a slice of dynamically unmarshalled json, so it needs to be treated properly.
+//
+// If a problem happens during execution, first argument is nil and second argument holds the error message.
+func GetImages(args ...string) ([]map[string]interface{}, error) {
+	args = append([]string{"images", "--format", "json"}, args...)
+	output, err := CmdOutput(args...)
+	if err != nil {
+		return nil, err
+	}
+
+	var images []map[string]interface{}
+
+	err = json.Unmarshal(output, &images)
+	if err != nil {
+		return nil, err
+	}
+
+	return images, nil
+}
+
+// ImageExists checks using Podman if an image with given ID/name exists.
+//
+// Parameter image is a name or an id of an image.
+func ImageExists(image string) bool {
+	args := []string{"image", "exists", image}
+
+	err := CmdRun(args...)
+	if err != nil {
+		return false
+	}
+
+	return true
+}
+
+// ContainerExists checks using Podman if a container with given ID/name exists.
+//
+// Parameter container is a name or an id of a container.
+func ContainerExists(containerName string) bool {
+	args := []string{"container", "exists", containerName}
+
+	err := CmdRun(args...)
+	if err != nil {
+		return false
+	}
+
+	return true
+}
+
+// PullImage pulls an image
+func PullImage(imageName string) error {
+	args := []string{"pull", imageName}
+
+	logLevel := fmt.Sprint(logrus.GetLevel())
+	if viper.GetBool("log-podman") {
+		args = append([]string{"--log-level", logLevel}, args...)
+	}
+
+	cmd := exec.Command("podman", args...)
+
+	var stderr strings.Builder
+	cmd.Stderr = &stderr
+
+	err := cmd.Run()
+	if logLevel == "debug" || logLevel == "trace" {
+		fmt.Fprint(os.Stderr, stderr.String())
+	}
+
+	if err != nil {
+		if strings.Contains(stderr.String(), "invalid status code from registry 503 (Service Unavailable)") || strings.Contains(stderr.String(), "received unexpected HTTP status: 503 Service Temporarily Unavailable") {
+			return ErrServiceUnavailable
```

Comment:
We shouldn't rely on looking for sub-strings in the error messages, unless these strings have been blessed as API. If these are just *informal* user-facing strings then they might get tweaked in a future release and break our error handling.

Instead, we should [add a specific exit code](https://github.com/containers/libpod/issues/6190) to `podman pull` to denote network failures, which can then be considered as API.

It's good to be paranoid about these error handling code paths. They don't get exercised under normal use, and sometimes it's not trivial to trigger these code paths even just for testing. So, if things silently break, then we are less likely to notice, and users who face the error in the wild would have even less of a clue.

Reply:
Yes, the sub-string as an error is not the most optimal way of handling this. I filed some time ago an issue that pointed out this behaviour but it was closed: https://github.com/containers/libpod/issues/5288.